From be032dd7499b81001f287eaca416aed6e7def371 Mon Sep 17 00:00:00 2001
From: David Sniatkiwicz <davidsn@marvell.com>
Date: Mon, 23 Jan 2017 11:04:23 +0200
Subject: [PATCH 0765/1345] fix: thermal: fix negative temperature calculation

commit  fda604d8d94cde0a22ad0d310278491e9922fb85 from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

negative temperature value was wrong since variables
were represented as 'unsigned' in the calculation.

Change-Id: I228387f98da065726d6d6f9d090a751e15896c49
Signed-off-by: David Sniatkiwicz <davidsn@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/35924
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Omri Itach <omrii@marvell.com>
Reviewed-by: Haim Boot <hayim@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/36138
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/thermal/armada_thermal.c |   11 ++++++-----
 1 file changed, 6 insertions(+), 5 deletions(-)

diff --git a/drivers/thermal/armada_thermal.c b/drivers/thermal/armada_thermal.c
index bdd2445..6d587cd 100644
--- a/drivers/thermal/armada_thermal.c
+++ b/drivers/thermal/armada_thermal.c
@@ -430,7 +430,7 @@ static int armada_get_temp(struct thermal_zone_device *thermal,
 {
 	struct armada_thermal_priv *priv = thermal->devdata;
 	unsigned long reg;
-	unsigned long m, b, div;
+	long m, b, div;
 
 	/* Valid check */
 	if (priv->data->is_valid && !priv->data->is_valid(priv)) {
@@ -448,9 +448,10 @@ static int armada_get_temp(struct thermal_zone_device *thermal,
 	div = priv->data->coef_div;
 
 	if (priv->data->inverted)
-		*temp = ((m * reg) - b) / div;
+		*temp = ((m * (long)reg) - b) / div;
 	else
-		*temp = (b - (m * reg)) / div;
+		*temp = (b - (m * (long)reg)) / div;
+
 	return 0;
 }
 
@@ -490,7 +491,7 @@ static int armada_cp110_get_temp(struct thermal_zone_device *thermal, int *temp)
 {
 	struct armada_thermal_priv *priv = thermal->devdata;
 	unsigned long reg;
-	unsigned long m, b, div;
+	long m, b, div;
 
 	/* Valid check */
 	if (priv->data->is_valid && !priv->data->is_valid(priv)) {
@@ -507,7 +508,7 @@ static int armada_cp110_get_temp(struct thermal_zone_device *thermal, int *temp)
 	m = priv->data->coef_m;
 	div = priv->data->coef_div;
 
-	*temp = ((m * reg) - b) / div;
+	*temp = ((m * (long)reg) - b) / div;
 
 	return 0;
 }
-- 
1.7.9.5

